File: java/arcs/core/crdt/CrdtCount.kt
/*
 * Copyright 2020 Google LLC.
 *
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 *
 * Code distributed by Google as part of this project is also subject to an additional IP rights
 * grant found at
 * http://polymer.github.io/PATENTS.txt
 */

package arcs.core.crdt

/** A [CrdtModel] capable of managing an increasing [Int] value. */
class CrdtCount : CrdtModel<CrdtCount.Data, CrdtCount.Operation, Int> {
  private var _data: Data = Data()
  override val versionMap: VersionMap
    get() = _data.versionMap.copy()
  override val data: Data
    get() = _data.copy(values = HashMap(_data.values), versionMap = _data.versionMap.copy())

  override val consumerView: Int
    get() = _data.values.values.sum()

  fun forActor(actor: Actor) = Scoped(actor)

  override fun merge(other: Data): MergeChanges<Data, Operation> {
    val myChanges = CrdtChange.Operations<Data, Operation>()
    val otherChanges = CrdtChange.Operations<Data, Operation>()

    other.values.forEach { (actor, otherValue) ->
      val myValue = _data.values[actor] ?: 0
      val myVersion = _data.versionMap[actor]
      val otherVersion = other.versionMap[actor]

      if (myValue > otherValue) {
        if (otherVersion >= myVersion) {
          throw CrdtException("Divergent versions encountered when merging CrdtCount models")
        }

        otherChanges +=
          Operation.MultiIncrement(actor, VersionMap(actor to myVersion), myValue - otherValue)
      } else if (otherValue > myValue) {
        if (myVersion >= otherVersion) {
          throw CrdtException("Divergent versions encountered when merging CrdtCount models")
        }

        myChanges +=
          Operation.MultiIncrement(actor, VersionMap(actor to otherVersion), otherValue - myValue)
        _data.values[actor] = otherValue
        _data.versionMap[actor] = otherVersion
      }
    }

    _data.values.forEach { (actor, myValue) ->
      if (actor in other.values) return@forEach
      if (actor in other.versionMap) {
        throw CrdtException("CrdtCount model has version but no value for actor: $actor")
      }

      otherChanges +=
        Operation.MultiIncrement(actor, VersionMap(actor to _data.versionMap[actor]), myValue)
    }
    return MergeChanges(myChanges, otherChanges)
  }

  override fun applyOperation(op: Operation): Boolean {
    // Only accept an operation if it is immediately consecutive to the versionMap for that actor.
    val opVersion = op.versionMap[op.actor]
    if (opVersion != _data.versionMap[op.actor] + 1) return false

    val delta = when (op) {
      is Operation.Increment -> 1
      is Operation.MultiIncrement -> op.delta
    }
    _data.values[op.actor] = (_data.values[op.actor] ?: 0) + delta
    _data.versionMap[op.actor] = opVersion
    return true
  }

  /** Scoped access to this [CrdtCount], where all ops will be marked as performed by [actor]. */
  inner class Scoped(private val actor: Actor) {
    operator fun plusAssign(delta: Int) {
      // [versionMap] getter makes a copy of _data.versionMap, so it's safe to increment in place.
      applyOperation(
        Operation.MultiIncrement(actor, versionMap.increment(actor), delta)
      )
    }
  }

  /** Internal representation of the count information. */
  data class Data(
    /** Increments seen from each actor. */
    val values: MutableMap<Actor, Int> = mutableMapOf(),
    /** Versions at each actor. */
    override var versionMap: VersionMap = VersionMap()
  ) : CrdtData

  /** Operation which can be performed on a [CrdtCount]. */
  sealed class Operation(open val actor: Actor) : CrdtOperation {
    /**
     * Increments the value by 1 for the given [actor] when going from one version to another
     * for that actor.
     */
    data class Increment(
      override val actor: Actor,
      override val versionMap: VersionMap
    ) : Operation(actor)

    /**
     * Increments the value by some positive integer [delta] for the given [actor] when going
     * from one version to another for that actor.
     */
    data class MultiIncrement(
      override val actor: Actor,
      override val versionMap: VersionMap,
      val delta: Int
    ) : Operation(actor) {
      init {
        require(delta > 0)
      }
    }
  }
}

File: src/crdt/internal/crdt-count.ts
/**
 * @license
 * Copyright (c) 2019 Google Inc. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * Code distributed by Google as part of this project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */

import {VersionMap, CRDTChange, CRDTModel, CRDTError, ChangeType, CRDTTypeRecord} from './crdt.js';
import {Dictionary} from '../../utils/hot.js';

type RawCount = number;

export type CountData = {values: Dictionary<number>, version: VersionMap};

type VersionInfo = {from: number, to: number};

export enum CountOpTypes {Increment, MultiIncrement}
export type CountOperation = {type: CountOpTypes.MultiIncrement, value: number, actor: string, version: VersionInfo} |
                             {type: CountOpTypes.Increment, actor: string, version: VersionInfo};

export interface CRDTCountTypeRecord extends CRDTTypeRecord {
  data: CountData;
  operation: CountOperation;
  consumerType: RawCount;
}

type CountModel = CRDTModel<CRDTCountTypeRecord>;
type CountChange = CRDTChange<CRDTCountTypeRecord>;

export class CRDTCount implements CountModel {
  private model: CountData = {values: {}, version: {}};

  merge(other: CountData): {modelChange: CountChange, otherChange: CountChange} {
    const otherChanges: CountOperation[] = [];
    const thisChanges: CountOperation[] = [];

    for (const key of Object.keys(other.values)) {
      const thisValue = this.model.values[key] || 0;
      const otherValue = other.values[key] || 0;
      const thisVersion = this.model.version[key] || 0;
      const otherVersion = other.version[key] || 0;
      if (thisValue > otherValue) {
        if (otherVersion >= thisVersion) {
          throw new CRDTError('Divergent versions encountered when merging CRDTCount models');
        }
        otherChanges.push({type: CountOpTypes.MultiIncrement, value: thisValue - otherValue, actor: key,
                           version: {from: otherVersion, to: thisVersion}});
      } else if (otherValue > thisValue) {
        if (thisVersion >= otherVersion) {
          throw new CRDTError('Divergent versions encountered when merging CRDTCount models');
        }
        thisChanges.push({type: CountOpTypes.MultiIncrement, value: otherValue - thisValue, actor: key,
                          version: {from: thisVersion, to: otherVersion}});
        this.model.values[key] = otherValue;
        this.model.version[key] = otherVersion;
      }
    }

    for (const key of Object.keys(this.model.values)) {
      if (other.values[key]) {
        continue;
      }
      if (other.version[key]) {
        throw new CRDTError(`CRDTCount model has version but no value for key ${key}`);
      }
      otherChanges.push({type: CountOpTypes.MultiIncrement, value: this.model.values[key], actor: key,
                         version: {from: 0, to: this.model.version[key]}});
    }

    return {modelChange: {changeType: ChangeType.Operations, operations: thisChanges}, otherChange: {changeType: ChangeType.Operations, operations: otherChanges}};
  }

  applyOperation(op: CountOperation) {
    let value: number;
    if (op.version.from !== (this.model.version[op.actor] || 0)) {
      return false;
    }
    if (op.version.to <= op.version.from) {
      return false;
    }
    if (op.type === CountOpTypes.MultiIncrement) {
      if (op.value < 0) {
        return false;
      }
      value = (this.model.values[op.actor] || 0) + op.value;
    } else {
      value = (this.model.values[op.actor] || 0) + 1;
    }

    this.model.values[op.actor] = value;
    this.model.version[op.actor] = op.version.to;
    return true;
  }

  private cloneMap<V>(map: Dictionary<V>) {
    const result: Dictionary<V> = {};
    Object.keys(map).forEach(key => result[key] = map[key]);
    return result;
  }

  getData() {
    return {values: this.cloneMap(this.model.values), version: this.cloneMap(this.model.version)};
  }

  getParticleView() {
    return Object.values(this.model.values).reduce((prev, current) => prev + current, 0);
  }
}
